src/router.rs

```diff
@@ -1,38 +1,129 @@
 use std::collections::HashMap;
+use std::sync::Arc;
 
 use crate::{
     endpoint::{BoxedEndpoint, Endpoint},
     url_table::{RouteMatch, UrlTable},
+    Middleware,
 };
 
-pub(crate) struct Router<Data> {
+/// TODO: Document `Router`
+pub struct Router<Data> {
     table: UrlTable<Resource<Data>>,
+    middleware: Vec<Arc<dyn Middleware<Data> + Send + Sync>>,
+}
+
+pub(crate) struct RouteResult<'a, Data> {
+    pub(crate) endpoint: &'a BoxedEndpoint<Data>,
+    pub(crate) params: RouteMatch<'a>,
+    pub(crate) middleware: &'a [Arc<dyn Middleware<Data> + Send + Sync>],
+}
+
+impl<Data> Default for Router<Data> {
+    fn default() -> Router<Data> {
+        Router::new()
+    }
 }
 
 impl<Data> Router<Data> {
     pub(crate) fn new() -> Router<Data> {
         Router {
             table: UrlTable::new(),
+            middleware: Vec::new(),
         }
     }
 
-    pub(crate) fn at<'a>(&'a mut self, path: &'a str) -> &mut Resource<Data> {
-        self.table.setup(path)
+    /// Add a new resource at `path`, relative to this router.
+    ///
+    /// Middlewares added before will be applied to the resource.
+    pub fn at<'a>(&'a mut self, path: &'a str) -> ResourceHandle<'a, Data> {
+        let table = self.table.setup_table(path);
+        let middleware = &*self.middleware;
+        ResourceHandle { table, middleware }
+    }
+
+    /// Add `middleware` to this router.
+    pub fn middleware(&mut self, middleware: impl Middleware<Data> + 'static) -> &mut Self {
+        self.middleware.push(Arc::new(middleware));
```

Comment:
We should consider extending *all* reachable middleware collections when this is called, so that the order of `at` and `middleware` calls continues to not matter.